flang/lib/Lower/ReductionRuntime.cpp

```diff
@@ -714,6 +714,85 @@ mlir::Value Fortran::lower::genProduct(Fortran::lower::FirOpBuilder &builder,
 
   return builder.create<fir::CallOp>(loc, func, args).getResult(0);
 }
+
+mlir::Value Fortran::lower::genDotProduct(Fortran::lower::FirOpBuilder &builder,
+                                          mlir::Location loc,
+                                          mlir::Value vectorABox,
+                                          mlir::Value vectorBBox,
+                                          mlir::Value resultBox) {
+  mlir::FuncOp func;
+  auto ty = vectorABox.getType();
+  auto arrTy = fir::dyn_cast_ptrOrBoxEleTy(ty);
+  auto eleTy = arrTy.cast<fir::SequenceType>().getEleTy();
+
+  if (eleTy.isF32())
+    func =
+        Fortran::lower::getRuntimeFunc<mkRTKey(DotProductReal4)>(loc, builder);
+  else if (eleTy.isF64())
+    func =
+        Fortran::lower::getRuntimeFunc<mkRTKey(DotProductReal8)>(loc, builder);
+
+  // REQUIRES LONG_DOUBLE=80
```

Comment:
I got a bit confused here by https://github.com/flang-compiler/f18-llvm-project/blob/fir-dev/flang/runtime/dot-product.cpp#L141-L145. Should I just use `#if LONG_DOUBLE == 80`? But then I won't be able to test this code-path.

Reply:
No. Take a look at how we have "struct ForcedProductReal10", "struct ForcedProductReal16", "ForcedProductInteger16", "ForcedProductComplex10", and "ForcedProductComplex16" in ReductionRuntime.cpp. Also, see how those are used in genProduct() in the same file. That way, real*10, real*16, complex*10, complex*16, integer*16 will just work if/when they are supported by the target.

Reply:
Ah, thank you for pointing this out!

How can I test that it works? This is what I get for `real(kind=10)`:
```
'fir.convert' op invalid type conversion
```
That's on an X86-64 machine. I would hope that we can force the frontend to generate this code regardless of whether the host systems supports extended real numbers.

Reply:
You should not get a fir error. The code should compile and link if the runtime routine is present for the target (although you may not get correct answers if the width is not supported natively by the target), or you should get a link time error. We have some regression tests in flang/test/Lower/intrinsic-procedures.f90 that use these higher precision data types (i.e., sum_test4 and product_test4 use complex(10), and spacing_test2 uses real*10).  But we are just testing the fir in intrinsic-procedures.f90; not whether the code links and runs.

If you test something like real*16, you should get a link time error for the missing routine. For example, consider the following SUM code that operates on real*16. It compiles successfully but then you get a link time error.

program p
  real*16:: x(2)

  x(1) = 1.
  x(2) = 9.

  print *, sum(x) 

end

When I compile it, I get the expected link time error on X86-64:

sum16.f90.o: In function `_QQmain':
/local/home/mleair/fir-dev/tests/sum/<stdin>:26: undefined reference to `_FortranASumReal16'
collect2: error: ld returned 1 exit status

Reply:
Thank you for confirming this and for all the references. There was a bug in my implementation - I was using incorrect function type in ` ForcedDotProductReal{N}`. This is now fixed and tested!